File: 0155-min-stack/0155-min-stack.java
class MinStack {
    
    Stack<Integer> stack;
    Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<Integer>();        
        minStack = new Stack<Integer>();                
    }
    
    public void push(int val) {
        stack.push(val);        
        
        if(minStack.isEmpty() || minStack.peek() >= val){
            minStack.push(val);    
        }
    }
    
    public void pop() {
        int popValue = stack.peek();
        
        stack.pop();
        
        if(minStack.peek() == popValue)
            minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();        
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.tempStack = []
        self.minStack = []

    def push(self, val: int) -> None:
        self.stack.append(val)

        while self.minStack and self.minStack[-1] < val:
            self.tempStack.append(self.minStack.pop())

        self.minStack.append(val)

        while self.tempStack:
            self.minStack.append(self.tempStack.pop())        

        
    def pop(self) -> None:
        num = self.stack.pop()

        while self.minStack:
            if self.minStack[-1] == num:
                self.minStack.pop()
                break
            
            self.tempStack.append(self.minStack.pop())

        while self.tempStack:
            self.minStack.append(self.tempStack.pop())                
        

    def top(self) -> int:
        if self.stack:
            return self.stack[-1]        

    def getMin(self) -> int:
        if self.minStack:
            return self.minStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()